main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -316,6 +317,52 @@ object Safety {
 
   }
 
+  /**
+    * Performs basic checks on the type cast `cast`. Returns a list of safety errors if there are
+    * any impossible casts.
+    *
+    * No primitive type can be cast to a reference type and vice-versa.
+    *
+    * No Bool type can be cast to a non-Bool type  and vice-versa.
+    */
+  private def checkCastSafety(cast: Expression.Cast)(implicit flix: Flix): List[SafetyError] = {
+    val tpe1 = Type.eraseAliases(cast.exp.tpe).baseType
+    val tpe2 = cast.declaredType.map(Type.eraseAliases).map(_.baseType)
+
+    val primitives = {
+      Type.Unit :: Type.Bool :: Type.Char ::
+        Type.Float32 :: Type.Float64 :: Type.Int8 ::
+        Type.Int16 :: Type.Int32 :: Type.Int64 ::
+        Type.Str :: Type.BigInt :: Type.BigDecimal :: Nil
+    }.toSet
+
+    (tpe1, tpe2) match {
+
+      // Reference types
+      case (t1, Some(Type.Cst(TypeConstructor.Ref, _))) if primitives.contains(t1) =>
```

Comment:
Ah, sorry, but reference types I did not mean ` Ref`, but anything that is an object on the JVM.